compiler/rustc_smir/src/stable_mir/mod.rs

```diff
@@ -126,8 +126,8 @@ pub trait Context {
     fn trait_decl(&mut self, trait_def: &TraitDef) -> TraitDecl;
     fn all_trait_impls(&mut self) -> ImplTraitDecls;
     fn trait_impl(&mut self, trait_impl: &ImplDef) -> ImplTrait;
-    fn generics_of(&mut self, generic_def: &GenericDef) -> Generics;
-    fn predicates_of(&mut self, trait_def: &TraitDef) -> GenericPredicates;
+    fn generics_of(&mut self, def_id: rustc_span::def_id::DefId) -> Generics;
+    fn predicates_of(&mut self, def_id: rustc_span::def_id::DefId) -> GenericPredicates;
```

Comment:
Shouldn't this be using the "stable" `DefId` defined on line 32 of this file? Seems surprising to be leaking `DefId` in the stable API here.

Reply:
Or alternatively, take a `&CrateItem` like other "item-agnostic" fn above?

Reply:
Yes, I was trying to show Oli the problem the current approach has. We need a trait on smir thst each different def implement or we need to reuse some existing type or just pass smir defid